runtime/gc_structs/ClassFCCSlotIterator.hpp

```diff
@@ -0,0 +1,69 @@
+
+/*******************************************************************************
+ * Copyright (c) 1991, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup GC_Structs
+ */
+
+#if !defined(CLASSFCCSLOTITERATOR_HPP_)
+#define CLASSFCCSLOTITERATOR_HPP_
+
+#include "j9.h"
+#include "j9nonbuilder.h"
+
+/**
+ * Iterate through slots in the flattened class cache which may contain a
+ *  class reference.
+ *
+ * @see GC_ClassFCCSLotIterator
+ * @ingroup GC_Structs
+ */
+class GC_ClassFCCSlotIterator
+{
+private:
+    J9Class* _clazz;
+    UDATA _numberOfFlattenedFields;
+    UDATA _index;
+public:
+	GC_ClassFCCSlotIterator(J9Class *clazz) :
+        _clazz(clazz),
+        _numberOfFlattenedFields(0),
+        _index(0)
+	{
+        if (J9_ARE_ALL_BITS_SET(_clazz->romClass->modifiers, J9AccValueType) && (NULL != clazz->flattenedClassCache)) {
```

Comment:
This line is correct, however I am wondering can we save on rom class access...
@tajila Is  check `(NULL != clazz->flattenedClassCache)` sufficient alone or it must be accompanied with check class is VT class?

Reply:
The reason for `J9_ARE_ALL_BITS_SET(_clazz->romClass->modifiers, J9AccValueType)` is that if the value is not a VT, it may have bits in that slot, but not be NULL. 

Ex: This came to light as an issue with array classes, where flattenedClassCache was not null, but it had a value of 0x4. Trying to get an entry from this causes a segfault.

It's worth looking at a more optimal method though.

Reply:
Proposed code is correct and I like it - it covers exact situation accurately. Reading of rom class also should not be a problem in this case due previous reads in this iterator.
I was wondering this slot is using exclusively for VT classes so check might be not necessary. If it is not a case, so keep existing solution

Reply:
Generally, if you have the ramClass you dont need to fetch to the romClass to check if its a VT, just use `J9_IS_J9CLASS_VALUETYPE(clazz)`.

For this, I think you need to test that its not an array and the FCC is not null. Non-VTs may have classes in the FCC.

```
if (J9_ARE_NO_BITS_SET(J9CLASS_FLAGS(clazz), J9AccClassArray) && (NULL != clazz->flattenedClassCache)) {
```